### PR TITLE
Allow module variables to pass validation

### PR DESCRIPTION
**Pull request by jbardin**
Validate module variables to the best of our ability during validation.

This challenge here is that modules are not expanded during validation, therefor variables that contain count or each references cannot be evaluated. An attempt at making the `Expander` handle unknown module expansions was started in #24586, but that was turning into too much of a re-write. After some testing, we may return to that with a simpler approach, but this allows terraform to work while missing some validations until plan time.

What is missing here, is that we cannot validate if an incorrect count/each reference was used in a variable, and we can't run variable validation functions with count/each references even if they are static. Each of these will be caught during plan, so there is no safety issue, just a matter of static validation that is possible, but not functioning. 

This also contains the renaming of the `evaluateResource*` functions to remove "Resource" since they are also used with modules. The `*Known` variants were replaced with `Value` versions that leverage the ability of `cty.Values` to indicate being unknown, in preparation for possibly making the instance expander handle those situations for us.

### CODE REVIEW

File: terraform/eval_count.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/zclconf/go-cty/cty/gocty"
 )
 
-// evaluateResourceCountExpression is our standard mechanism for interpreting an
+// evaluateCountExpression is our standard mechanism for interpreting an
```
Comment (mildwonkey) [inline terraform/eval_count.go:14]: You may want to update the comment so it's less resource-specific!

Comment (jbardin) [reply terraform/eval_count.go:14]: Thanks! I did! it was just in the other PR :(

File: terraform/eval_count.go [hunk 2]
```diff
@@ -40,22 +35,29 @@ func evaluateResourceCountExpression(expr hcl.Expression, ctx EvalContext) (int,
...
+	count, _ := countVal.AsBigFloat().Int64()
+	return int(count), diags
 }
 
-// evaluateResourceCountExpressionKnown is like evaluateResourceCountExpression
-// except that it handles an unknown result by returning count = 0 and
-// a known = false, rather than by reporting the unknown value as an error
-// diagnostic.
-func evaluateResourceCountExpressionKnown(expr hcl.Expression, ctx EvalContext) (count int, known bool, diags tfdiags.Diagnostics) {
+// evaluateCountExpressionValue is like evaluateCountExpression
```
Comment (pselle) [inline terraform/eval_count.go:47]: I like this Known => Value change, and killing the bool

File: terraform/eval_for_each.go [hunk 3]
```diff
@@ -50,44 +61,42 @@ func evaluateResourceForEachExpressionKnown(expr hcl.Expression, ctx EvalContext
...
-	}
+		return nullMap, diags
 
-	// If the map is empty ({}), return an empty map, because cty will return nil when representing {} AsValueMap
-	// This also covers an empty set (toset([]))
-	if forEachVal.LengthInt() == 0 {
-		return map[string]cty.Value{}, true, diags
+	case forEachVal.LengthInt() == 0:
+		// If the map is empty ({}), return an empty map, because cty will
+		// return nil when representing {} AsValueMap This also covers an empty
```
Comment (pselle) [inline terraform/eval_for_each.go:80]: Soooo minor, but moving this comment needs a little grammar fix
```suggestion
		// return nil when representing {} AsValueMap. This also covers an empty
```

File: terraform/eval_for_each_test.go [hunk 4]
```diff
@@ -159,23 +159,15 @@ func TestEvaluateResourceForEachExpressionKnown(t *testing.T) {
...
 				t.Errorf("unexpected diagnostics %s", spew.Sdump(diags))
 			}
 
-			if known {
-				t.Errorf("got %v known, want false", known)
+			if forEachVal.IsKnown() {
+				t.Error("got known, want unknown")
 			}
-
-			if len(forEachMap) != 0 {
```
Comment (pselle) [inline terraform/eval_for_each_test.go:172]: What's going on with this deletion? Is this bit of the test not useful, or what's up?

File: terraform/eval_variable.go [hunk 5]
```diff
@@ -42,12 +43,12 @@ type EvalModuleCallArgument struct {
 	Expr           hcl.Expression
 	ModuleInstance addrs.ModuleInstance
 
-	// If this flag is set, any diagnostics are discarded and this operation
-	// will always succeed, though may produce an unknown value in the
-	// event of an error.
-	IgnoreDiagnostics bool
```
Comment (pselle) [inline terraform/eval_variable.go:48]: Nice! I've thought this looked like cruft

File: terraform/eval_for_each_test.go (hunk 4 above)
Comment (jbardin) [reply terraform/eval_for_each_test.go:172]: The result is now a `cty.Value`, so there is no length if the value is null or unknown, and we know it's unknown from the check above.

**Pull request merged**